src/Containers/Microsoft.NET.Build.Containers/SourceImageReference.cs

```diff
@@ -4,20 +4,25 @@
 namespace Microsoft.NET.Build.Containers;
 
 /// <summary>
-/// Represents a reference to a Docker image. A reference is made of a registry, a repository (aka the image name) and a tag.
+/// Represents a reference to a Docker image. A reference is made of a registry, a repository (aka the image name) and a tag or digest.
 /// </summary>
-internal readonly record struct SourceImageReference(Registry? Registry, string Repository, string Tag)
+internal readonly record struct SourceImageReference(Registry? Registry, string Repository, string Tag, string? Digest)
 {
     public override string ToString()
     {
-        if (Registry is { } reg) 
+        string sourceImageReference = RepositoryAndTag;
```

Comment:
It may be confusing to have `latest` appear as a tag in `ToString` when a user specifies only a digest.

To avoid that, we could do the defaulting to `latest` through a property of `SourceImageReference`:

```cs
internal readonly record struct SourceImageReference
{
    public string Reference
          => !string.IsNullOrEmpty(Digest) ? Digest : !string.IsNullOrEmpty(Tag) ? Tag : "latest";
}
```

Reply:
Can you elaborate/expand? I don't quite understand. 🫣 

Are you looking to avoid something like: `mcr.microsoft.com/dotnet/aspnet:latest@sha256:12345....` ?

But `mcr.microsoft.com/dotnet/aspnet:8.0@sha256:12345...` is OK?

Reply:
If the user specified `mcr.microsoft.com/dotnet/aspnet@sha256:12345....`, it shouldn't appear as `mcr.microsoft.com/dotnet/aspnet:latest@sha256:12345....` for example when logged here:
https://github.com/dotnet/sdk/blob/f9c3951b409a1011ef0f81cbca479a6535e96a61/src/Containers/Microsoft.NET.Build.Containers/ContainerBuilder.cs#L100

Reply:
Wouldn't your suggestion [here](https://github.com/dotnet/sdk/pull/44461#discussion_r1818676639) always replace the tag with digest (if provided)?

Meaning

`mcr.microsoft.com/dotnet/aspnet:8.0@sha256:12345...`

would become

`mcr.microsoft.com/dotnet/aspnet@sha256:12345...`

Reply:
I think this might be impacted a bit by the decision to flow in `latest` from ParseContainerProperties - IIRC if no tag is specified then the [null-coalesce happening in that Task](https://github.com/dotnet/sdk/pull/44461/files#diff-8197d462161e626a4d996dc6297e2e0c02bb8cce1d2c4d14c07724acf5a02a92R170) would default the label to `"latest"`. I think this might actually be buggy - we should probably not do the defaulting during `ParseContainerProperties` because a user-provided reference like `mcr.microsoft/dotnet/aspnet@sha256:deadbeefdeadbeefdeadbeefdeadbeef` _doesn't_ mean `latest`. 

So I think
* ParseContainerProperties should not default Tags to latest
* We should accept empty/null ParsedContainerTag values out of ParseContainerProperties
* CreateNewImage (maybe inside the SourceImageReference, maybe in some other location) should perform some logic on the potentially-null-or-empty Tag like so
  * null/empty Tag, non-null/empty Digest -> the reference is to a digest
  * non-null/empty Tag, null/empty Digest -> the reference is to a tag
  * non-null/empty Tag, non-null/empty Digest -> the reference is to a digest, but informationally a digest that at one point the tag pointed to

Reply:
> maybe inside the SourceImageReference

This is what I am suggesting (https://github.com/dotnet/sdk/pull/44461#discussion_r1818676639).

Reply:
agree @tmds - but what use is a PM if I can't write 200 words when 20 would do? 😆

Reply:
Pushed an update. Hope this is what you were looking for. 😅

Reply:
ping @baronfel @tmds ☝️ 🙏

Reply:
Yes, this is the behavior we want.